### PR TITLE
LLC service version

### PR DESCRIPTION
**Pull request by qwordy**
Resolve https://github.com/Azure/autorest.java/issues/1110
`ClientBuilder` class will provide a method to set service version. It maps to `api-version` in swagger.
```
    public XXXClientBuilder serviceVersion(XXXServiceVersion version) {
        this.version = version;
        return this;
    }
```
`XXXServiceVersion` is an enum type. It defines service versions supported by service and latest version.
```
public enum XXXServiceVersion implements ServiceVersion {
    V2020_06_12("2020-06-12"),
    V2020_08_04("2020-08-04"),
    V2020_10_02("2020-10-02");

    private final String version;

    XXXServiceVersion(String version) {
        this.version = version;
    }

    @Override
    public String getVersion() {
        return this.version;
    }

    public static XXXVersion getLatest() {
        return V2020_10_02;
    }
}
```
Service version list is configured at config file or command parameter.
Config code is in yaml syntax. They are sorted in an ascending order. The bottom one is latest service version. Please add quotes around date.
```
service-versions:
  - "2020-06-12"
  - "2020-08-04"
  - "2020-10-02"
```
`ClientBuilder` class has
```
private XXXServiceVersion serviceVersion;

public XXXClientBuilder serviceVersion(XXXServiceVersion serviceVersion) {
    this.serviceVersion = serviceVersion;
    return this;
}

private XXXClientImpl buildInnerClient() {
    if (serviceVersion == null) {
        this.serviceVersion = XXXServiceVersion.getLatest();
    }
    ...
    XXXClientImpl client =
            new XXXClientImpl(pipeline, serializerAdapter, endpoint, serviceVersion.getVersion());
    ...
```
If user does not set service version, it will use latest service version by default.
`ClientImpl` class uses String type `apiVersion`.
```
private final String apiVersion;
```

### CODE REVIEW

File: extension-base/src/main/java/com/azure/autorest/extension/base/plugin/NewPlugin.java [hunk 1]
```diff
@@ -80,6 +80,10 @@ public boolean getBooleanValue(String key, boolean defaultValue) {
         }
     }
 
+    public List<String> getListValue(String key) {
```
Comment (weidongxu-microsoft) [inline extension-base/src/main/java/com/azure/autorest/extension/base/plugin/NewPlugin.java:83]: Might not worth adding to `NewPlugin` if it only works for `String`. If you want to put it here, maybe make it more generic by using `List<T>`? Or just have the method as private in `JavaSettings`.

File: extension-base/src/main/java/com/azure/autorest/extension/base/plugin/JavaSettings.java [hunk 2]
```diff
@@ -545,6 +548,12 @@ public boolean shouldUseIterable() {
         return useIterable;
     }
 
+    private final List<String> serviceVersions;
+
+    public List<String> getServiceVersions() {
```
Comment (weidongxu-microsoft) [inline extension-base/src/main/java/com/azure/autorest/extension/base/plugin/JavaSettings.java:553]: Worth add some javadoc.

File: javagen/src/main/java/com/azure/autorest/template/ServiceClientBuilderTemplate.java [hunk 3]
```diff
@@ -122,9 +122,29 @@ public final void write(ServiceClient serviceClient, JavaFile javaFile) {
                     javaBlock.line("this.pipelinePolicies = new ArrayList<>();");
                 });
             }
+
+            // ServiceVersion in builder
+            List<ServiceClientProperty> serviceClientProperties = new ArrayList<>(serviceClient.getProperties());
+            if (settings.isLowLevelClient()) {
+                for (int i = 0; i < serviceClientProperties.size(); i++) {
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ServiceClientBuilderTemplate.java:129]: nit. Might worth use `ListIterator`, so you can replace the value in-place (without the `set(i, ...)`).

File: javagen/src/main/java/com/azure/autorest/template/ServiceVersionTemplate.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    @Override
+    public void write(ServiceClient model, JavaFile javaFile) {
+        // imports
+        Set<String> imports = new HashSet<>();
+        imports.add("com.azure.core.util.ServiceVersion");
+        javaFile.declareImport(imports);
+
+        javaFile.javadocComment(comment -> {
+            comment.description("Service version");
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ServiceVersionTemplate.java:29]: Could be improved to mention about the service?

File: javagen/src/main/java/com/azure/autorest/template/ServiceClientBuilderTemplate.java [hunk 5]
```diff
@@ -122,9 +122,29 @@ public final void write(ServiceClient serviceClient, JavaFile javaFile) {
                     javaBlock.line("this.pipelinePolicies = new ArrayList<>();");
                 });
             }
+
+            // ServiceVersion in builder
+            List<ServiceClientProperty> serviceClientProperties = new ArrayList<>(serviceClient.getProperties());
```
Comment (jianghaolu) [inline javagen/src/main/java/com/azure/autorest/template/ServiceClientBuilderTemplate.java:127]: Should this logic be moved to ServiceClientMapper?

File: extension-base/src/main/java/com/azure/autorest/extension/base/plugin/JavaSettings.java (hunk 2 above)
Comment (qwordy) [reply extension-base/src/main/java/com/azure/autorest/extension/base/plugin/JavaSettings.java:553]: Done

File: extension-base/src/main/java/com/azure/autorest/extension/base/plugin/NewPlugin.java (hunk 1 above)
Comment (qwordy) [reply extension-base/src/main/java/com/azure/autorest/extension/base/plugin/NewPlugin.java]: Change to inline code now

File: javagen/src/main/java/com/azure/autorest/template/ServiceClientBuilderTemplate.java (hunk 5 above)
Comment (qwordy) [reply javagen/src/main/java/com/azure/autorest/template/ServiceClientBuilderTemplate.java:127]: The properties in service client are also used in `ClientImpl`. I need `apiVersion` in `ClientImpl` and `serviceVersion` in `ClientBuilder`. I create a new list here. `serviceClient` remains unmodified. We can also add `builderProperties` in `serviceClient`, then template code can be simplified.

File: javagen/src/main/java/com/azure/autorest/template/ServiceClientBuilderTemplate.java (hunk 3 above)
Comment (qwordy) [reply javagen/src/main/java/com/azure/autorest/template/ServiceClientBuilderTemplate.java:129]: https://github.com/Azure/autorest.java/pull/1126#discussion_r681551103

**Review by weidongxu-microsoft — APPROVED** [review]

File: javagen/src/main/java/com/azure/autorest/template/ServiceClientBuilderTemplate.java (hunk 5 above)
Comment (jianghaolu) [reply javagen/src/main/java/com/azure/autorest/template/ServiceClientBuilderTemplate.java:127]: Is there a reason you need to keep it as `apiVersion` in `ClientImpl`? I know we send it as `api-version` on the wire but locally we can name everything as `serviceVersion` right?

Comment (qwordy) [reply javagen/src/main/java/com/azure/autorest/template/ServiceClientBuilderTemplate.java:127]: I changed `apiVersion` to `serviceVersion` in `ClientImpl` and `ClientBuilder`. Proxy method keeps `apiVersion`.

File: extension-base/src/main/java/com/azure/autorest/extension/base/plugin/JavaSettings.java (hunk 2 above)
Comment (qwordy) [reply extension-base/src/main/java/com/azure/autorest/extension/base/plugin/JavaSettings.java:557]: Added

File: javagen/src/main/java/com/azure/autorest/template/ServiceClientBuilderTemplate.java (hunk 3 above)
Comment (qwordy) [reply javagen/src/main/java/com/azure/autorest/template/ServiceClientBuilderTemplate.java]: This code snippet is removed.

File: javagen/src/main/java/com/azure/autorest/template/ServiceVersionTemplate.java (hunk 4 above)
Comment (qwordy) [reply javagen/src/main/java/com/azure/autorest/template/ServiceVersionTemplate.java]: Added

**Pull request merged**